src/cmap/connection_pool.ts

```diff
@@ -642,44 +653,49 @@ function processWaitQueue(pool: ConnectionPool) {
     }
   }
 
-  const maxPoolSize = pool.options.maxPoolSize;
-  if (pool.waitQueueSize && (maxPoolSize <= 0 || pool.totalConnectionCount < maxPoolSize)) {
+  const { maxPoolSize, maxConnecting } = pool.options;
+  while (
+    pool.waitQueueSize &&
+    pool.pendingConnectionCount < maxConnecting &&
+    pool.availableConnectionCount <= 0 &&
+    (maxPoolSize <= 0 || pool.totalConnectionCount < maxPoolSize)
```

Comment:
When does maxPoolSize go negative? should this be === 0?

Reply:
this is just a holdover of the original logic (see L646 of the original code)